OpenIGTLinkIF/MRML/vtkMRMLIGTLConnectorNode.cxx

```diff
@@ -353,12 +366,76 @@ void vtkMRMLIGTLConnectorNode::vtkInternal::ProcessIncomingDeviceModifiedEvent(
+        imageMetaNode->ClearImageMetaElement();
+        igtlioImageMetaConverter::ImageMetaDataList imageMetaList = imageMetaDevice->GetContent().ImageMetaDataElements;
+        for (igtlioImageMetaConverter::ImageMetaDataList::iterator imageMetaIt = imageMetaList.begin(); imageMetaIt != imageMetaList.end(); ++imageMetaIt)
+        {
+          vtkMRMLImageMetaElement imageMetaElement;
+          imageMetaElement.DeviceName = imageMetaIt->DeviceName;
+          imageMetaElement.Modality = imageMetaIt->Modality;
+          imageMetaElement.Name = imageMetaIt->Name;
+          imageMetaElement.PatientID = imageMetaIt->PatientID;
+          imageMetaElement.PatientName = imageMetaIt->PatientName;
+          imageMetaElement.ScalarType = imageMetaIt->ScalarType;
+          for (int i = 0; i < 3; ++i)
+          {
+            imageMetaElement.Size[i] = imageMetaIt->Size[i];
+          }
+          imageMetaElement.TimeStamp = imageMetaIt->Timestamp;
+          imageMetaNode->AddImageMetaElement(imageMetaElement);
+        }
+      }
+    }
+    else if (strcmp(deviceType.c_str(), "LBMETA") == 0)
+    {
+      igtlioLabelMetaDevice* labelMetaDevice = reinterpret_cast<igtlioLabelMetaDevice*>(modifiedDevice);
+      if (strcmp(modifiedNode->GetName(), deviceName.c_str()) == 0)
+      {
+        vtkMRMLLabelMetaListNode* labelMetaNode = vtkMRMLLabelMetaListNode::SafeDownCast(modifiedNode);
+        labelMetaNode->ClearLabelMetaElement();
+        igtlioLabelMetaConverter::LabelMetaDataList labelMetaList = labelMetaDevice->GetContent().LabelMetaDataElements;
+        for (igtlioLabelMetaConverter::LabelMetaDataList::iterator labelMetaIt = labelMetaList.begin(); labelMetaIt != labelMetaList.end(); ++labelMetaIt)
+        {
+          vtkMRMLLabelMetaListNode::LabelMetaElement labelMetaElement;
+          labelMetaElement.DeviceName = labelMetaIt->DeviceName;
+          labelMetaElement.Label = labelMetaIt->Label;
+          labelMetaElement.Name = labelMetaIt->Name;
+          labelMetaElement.Owner = labelMetaIt->Owner;
+          for (int i = 0; i < 4; ++i)
+          {
+            labelMetaElement.RGBA[i] = labelMetaIt->RGBA[i];
+          }
+          for (int i = 0; i < 3; ++i)
+          {
+            labelMetaElement.Size[i] = labelMetaIt->Size[i];
+          }
+          labelMetaNode->AddLabelMetaElement(labelMetaElement);
+        }
+      }
     }
   }
 
+  vtkMRMLIGTLQueryNode* queryNode = this->GetQueryNodeForDevice(modifiedDevice);
+  if (queryNode && modifiedNode)
+  {
+    queryResponses.push_back(queryNode);
+    queryNode->SetResponseDataNodeID(modifiedNode->GetID());
+  }
+  this->RemovePendingQuerynode(queryNode);
+  for (QueryListType::iterator queryResponseIt = queryResponses.begin(); queryResponseIt != queryResponses.end(); ++queryResponseIt)
+  {
+    vtkMRMLIGTLQueryNode* queryNode = *queryResponseIt;
+    queryNode->SetQueryStatus(vtkMRMLIGTLQueryNode::STATUS_SUCCESS);
```

Comment:
Can a query time out? I think timeout feature existed before but I don't see it now.

Reply:
Ah, I meant to add the time out as well.